File: solutions/Djikstra/dijkstra.rb
#!/usr/bin/env ruby



def mul(a, b)
  mul_table = [[0,  0,  0,  0,  0],
               [0,  1,  2,  3,  4],
               [0,  2, -1,  4, -3],
               [0,  3, -4, -1,  2],
               [0,  4,  3, -2, -1]]
  sign = a * b > 0 ? 1 : -1
  return sign * mul_table[a.abs][b.abs]
end

def reduce(codes, times)
  value = codes.inject(1){ |value, code| mul(value, code) }
  return exponentiate(value, times)
end

def exponentiate(value, times)
  limit = times % 4
  power = 1
  while limit > 0 do
    power = mul(power, value)
    limit -= 1
  end
  return power
end

def viable_split(codes, times)
  maybe2 = 1
  maybe3 = 1
  while times > 0 do
    codes.map{|code|
      if maybe2 != 2 then
        maybe2 = mul(maybe2, code)
      elsif maybe3 != 3 then
        maybe3 = mul(maybe3, code)
      else
        return true
      end
    }
    times -= 1
  end
  return maybe2 == 2 && maybe3 == 3
end


#x = gets
#y = gets
#puts "x = #{x}"
#puts "y = #{y}"

test_cases_str = gets
test_cases = test_cases_str.to_i

for test in 1..test_cases do
  inputs_str = gets
  inputs = inputs_str.split.map{|str| str.to_i}
  test_L = inputs[0]
  test_X = inputs[1]
  inputs_str = gets
  test_I = inputs_str.split[0].chars.map{|ch| ch.ord - "i".ord + 2}

  test1 = reduce(test_I, test_X) == -1
  test2 = viable_split(test_I, [8,test_X].min)
  puts "Case ##{test}: #{test1 && test2 ? "YES" : "NO"}\n"
end

File: solutions/Djikstra/dijkstra.go
package main

import "fmt"
import (
	"math"
	"strconv"
)

var mul_table = [5][5]int{
	{0,  0,  0,  0,  0},
	{0,  1,  2,  3,  4},
	{0,  2, -1,  4, -3},
	{0,  3, -4, -1,  2},
	{0,  4,  3, -2, -1},
}

func mul(a, b int) int {
	var sign int
	if a*b > 0 {
		sign = 1
	} else {
		sign = -1
	}
	var row = int(math.Abs(float64(a)))
	var col = int(math.Abs(float64(b)))
	return sign * mul_table[row][col]
}

func reduce(codes []int, times int) int {
	var value = 1
	for _, code := range codes {
		value = mul(value, code)
	}
	return exponentiate(value, times)
}

func exponentiate(value int, times int) int {
	var limit = times % 4
	var power = 1
	for limit > 0 {
		power = mul(power, value)
		limit--
	}
	return power
}

func viable_split(codes []int, times int) bool {
	var maybe_2, maybe_3 int = 1, 1
	for times > 0 {
		for _, code := range codes {
			if maybe_2 != 2 {
				maybe_2 = mul(maybe_2, code)
			} else if maybe_3 != 3 {
				maybe_3 = mul(maybe_3, code)
			} else {
				return true
			}
		}
		times--
	}
	return maybe_2 == 2 && maybe_3 == 3
}

func main() {
	var test_cases_str string
	fmt.Scan(&test_cases_str)
	test_cases, _ := strconv.Atoi(test_cases_str)

	var test_L_str, test_X_str, test_I_str string
	var test_I []int

	for test := 1; test <= test_cases; test++ {
		fmt.Scan(&test_L_str, &test_X_str, &test_I_str)

		test_L, _ := strconv.Atoi(test_L_str)
		test_X, _ := strconv.Atoi(test_X_str)

		test_I = make([]int, test_L)
		for idx, ch := range test_I_str {
			test_I[idx] = int(ch) - int('i') + 2
		}

		test_1 := reduce(test_I, test_X) == -1
		test_2 := viable_split(test_I, int(math.Min(8, float64(test_X))))

		if test_1 && test_2 {
			fmt.Printf("Case #%d: YES\n", test)
		} else {
			fmt.Printf("Case #%d: NO\n", test)
		}
	}
}
